File: backend/app/services/translation_service.py
from sqlalchemy.orm import Session
from typing import List, Optional
from app.models.translation_set import TranslationSet
import logging

logger = logging.getLogger(__name__)

class TranslationService:
    def __init__(self):
        pass

    def create_translation(self, db: Session, content_id: str, language_code: str,
                          translated_title: str, translated_content: str,
                          summary: Optional[str] = None) -> TranslationSet:
        """Create a new translation for content."""
        translation = TranslationSet(
            content_id=content_id,
            language_code=language_code,
            translated_title=translated_title,
            translated_content=translated_content,
            summary=summary
        )
        db.add(translation)
        db.commit()
        db.refresh(translation)
        return translation

    def get_translation(self, db: Session, content_id: str, language_code: str) -> Optional[TranslationSet]:
        """Get a specific translation for content and language."""
        return db.query(TranslationSet).filter(
            TranslationSet.content_id == content_id,
            TranslationSet.language_code == language_code
        ).first()

    def get_translations_for_content(self, db: Session, content_id: str) -> List[TranslationSet]:
        """Get all translations for a specific content."""
        return db.query(TranslationSet).filter(
            TranslationSet.content_id == content_id
        ).all()

    def get_translations_by_language(self, db: Session, language_code: str) -> List[TranslationSet]:
        """Get all translations for a specific language."""
        return db.query(TranslationSet).filter(
            TranslationSet.language_code == language_code
        ).all()

    def update_translation(self, db: Session, content_id: str, language_code: str,
                          **kwargs) -> Optional[TranslationSet]:
        """Update an existing translation."""
        translation = db.query(TranslationSet).filter(
            TranslationSet.content_id == content_id,
            TranslationSet.language_code == language_code
        ).first()

        if translation:
            for key, value in kwargs.items():
                if hasattr(translation, key):
                    setattr(translation, key, value)
            db.commit()
            db.refresh(translation)

        return translation

    def delete_translation(self, db: Session, content_id: str, language_code: str) -> bool:
        """Delete a translation."""
        translation = db.query(TranslationSet).filter(
            TranslationSet.content_id == content_id,
            TranslationSet.language_code == language_code
        ).first()

        if translation:
            db.delete(translation)
            db.commit()
            return True

        return False

    def get_available_languages_for_content(self, db: Session, content_id: str) -> List[str]:
        """Get all available language codes for a specific content."""
        translations = db.query(TranslationSet.language_code).filter(
            TranslationSet.content_id == content_id
        ).all()

        return [t.language_code for t in translations]

    def get_translation_progress(self, db: Session, language_code: str) -> dict:
        """Get translation progress for a language."""
        # Get total book content
        from app.models.book_content import BookContent
        total_content = db.query(BookContent).count()

        # Get translated content for the specified language
        translated_content = db.query(TranslationSet).filter(
            TranslationSet.language_code == language_code
        ).count()

        # Calculate progress percentage
        progress_percentage = (translated_content / total_content * 100) if total_content > 0 else 0

        return {
            "language_code": language_code,
            "total_content": total_content,
            "translated_content": translated_content,
            "progress_percentage": progress_percentage
        }

    def validate_translation_content(self, content_id: str, language_code: str,
                                   translated_title: str, translated_content: str) -> dict:
        """Validate translation content quality and format."""
        errors = []

        # Check if required fields are provided
        if not translated_title or len(translated_title.strip()) == 0:
            errors.append("Translated title is required")

        if not translated_content or len(translated_content.strip()) == 0:
            errors.append("Translated content is required")

        # Check for quality issues
        if translated_content:
            # Check for repeated characters that might indicate encoding issues
            import re
            if re.search(r'(.)\1{5,}', translated_content):
                errors.append("Translation contains repeated characters that may indicate encoding issues")

            # Check for unusual character sequences
            if re.search(r'[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]', translated_content):
                errors.append("Translation contains control characters that may indicate processing errors")

        # Check if translation is too short compared to expected content length
        # This is a simple check - in a real implementation, we might compare to original content
        if translated_content and len(translated_content.strip()) < 10:
            errors.append("Translation content appears to be too short")

        return {
            "is_valid": len(errors) == 0,
            "errors": errors
        }

# Create a global instance
translation_service = TranslationService()

File: website/src/utils/translationService.js
/**
 * Service for handling translation functionality
 */

const API_BASE_URL = process.env.REACT_APP_API_URL || 'http://localhost:8000';

// Simple in-memory cache for translations
const translationCache = new Map();
const CACHE_DURATION = 5 * 60 * 1000; // 5 minutes

/**
 * Check if cached translation is still valid
 */
const isCacheValid = (cacheEntry) => {
  if (!cacheEntry) return false;
  const now = Date.now();
  return (now - cacheEntry.timestamp) < CACHE_DURATION;
};

/**
 * Fetch translation for specific content and language
 */
export const fetchTranslation = async (contentId, languageCode) => {
  const cacheKey = `${contentId}_${languageCode}`;

  // Check cache first
  const cached = translationCache.get(cacheKey);
  if (cached && isCacheValid(cached)) {
    return cached.data;
  }

  try {
    const response = await fetch(
      `${API_BASE_URL}/api/translation/${contentId}/${languageCode}`,
      {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    );

    if (!response.ok) {
      if (response.status === 404) {
        // Cache the "not found" result
        translationCache.set(cacheKey, {
          data: null,
          timestamp: Date.now()
        });
        return null; // Translation doesn't exist
      }
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    const data = await response.json();

    // Cache the successful result
    translationCache.set(cacheKey, {
      data: data,
      timestamp: Date.now()
    });

    return data;
  } catch (error) {
    console.error('Error fetching translation:', error);

    // If there was an error but we have a cached result, return it
    const fallback = translationCache.get(cacheKey);
    if (fallback && isCacheValid(fallback)) {
      return fallback.data;
    }

    throw error;
  }
};

/**
 * Fetch all translations for a specific content
 */
export const fetchTranslationsForContent = async (contentId) => {
  const cacheKey = `translations_${contentId}`;

  // Check cache first
  const cached = translationCache.get(cacheKey);
  if (cached && isCacheValid(cached)) {
    return cached.data;
  }

  try {
    const response = await fetch(
      `${API_BASE_URL}/api/translation/${contentId}`,
      {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    );

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    const data = await response.json();

    // Cache the result
    translationCache.set(cacheKey, {
      data: data,
      timestamp: Date.now()
    });

    return data;
  } catch (error) {
    console.error('Error fetching translations for content:', error);

    // If there was an error but we have a cached result, return it
    const fallback = translationCache.get(cacheKey);
    if (fallback && isCacheValid(fallback)) {
      return fallback.data;
    }

    throw error;
  }
};

/**
 * Get translation progress for a language
 */
export const getTranslationProgress = async (languageCode) => {
  const cacheKey = `progress_${languageCode}`;

  // Check cache first
  const cached = translationCache.get(cacheKey);
  if (cached && isCacheValid(cached)) {
    return cached.data;
  }

  try {
    const response = await fetch(
      `${API_BASE_URL}/api/translation/progress/${languageCode}`,
      {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    );

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    const data = await response.json();

    // Cache the result (with a longer cache duration for progress data)
    translationCache.set(cacheKey, {
      data: data,
      timestamp: Date.now()
    });

    return data;
  } catch (error) {
    console.error('Error fetching translation progress:', error);

    // If there was an error but we have a cached result, return it
    const fallback = translationCache.get(cacheKey);
    if (fallback && isCacheValid(fallback)) {
      return fallback.data;
    }

    throw error;
  }
};

/**
 * Create a new translation
 */
export const createTranslation = async (contentId, languageCode, translationData) => {
  try {
    const response = await fetch(
      `${API_BASE_URL}/api/translation/`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          content_id: contentId,
          language_code: languageCode,
          translated_title: translationData.translatedTitle,
          translated_content: translationData.translatedContent,
          summary: translationData.summary
        }),
      }
    );

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    const data = await response.json();

    // Clear related cache entries
    translationCache.delete(`${contentId}_${languageCode}`);
    translationCache.delete(`translations_${contentId}`);

    return data;
  } catch (error) {
    console.error('Error creating translation:', error);
    throw error;
  }
};

/**
 * Update an existing translation
 */
export const updateTranslation = async (contentId, languageCode, translationData) => {
  try {
    const response = await fetch(
      `${API_BASE_URL}/api/translation/${contentId}/${languageCode}`,
      {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          translated_title: translationData.translatedTitle,
          translated_content: translationData.translatedContent,
          summary: translationData.summary
        }),
      }
    );

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    const data = await response.json();

    // Clear related cache entries
    translationCache.delete(`${contentId}_${languageCode}`);
    translationCache.delete(`translations_${contentId}`);

    return data;
  } catch (error) {
    console.error('Error updating translation:', error);
    throw error;
  }
};

/**
 * Check if translation is available for content and language
 */
export const isTranslationAvailable = async (contentId, languageCode) => {
  try {
    const translation = await fetchTranslation(contentId, languageCode);
    return !!translation;
  } catch (error) {
    console.error('Error checking translation availability:', error);
    return false;
  }
};

/**
 * Validate translation quality and content
 */
export const validateTranslation = (translationData, originalContent = '') => {
  const errors = [];

  // Check if required fields exist
  if (!translationData.translated_title || translationData.translated_title.trim().length === 0) {
    errors.push('Translated title is required');
  }

  if (!translationData.translated_content || translationData.translated_content.trim().length === 0) {
    errors.push('Translated content is required');
  }

  // Check if translation is too short compared to original (may indicate poor quality)
  if (originalContent && translationData.translated_content) {
    const originalLength = originalContent.length;
    const translatedLength = translationData.translated_content.length;

    // If translated content is less than 30% of original, flag as potentially low quality
    if (originalLength > 0 && translatedLength / originalLength < 0.3) {
      errors.push('Translated content appears to be significantly shorter than original, which may indicate low quality');
    }
  }

  // Check for potential machine translation artifacts or quality issues
  if (translationData.translated_content) {
    // Check for repeated characters that might indicate encoding issues
    if (/(.)\1{5,}/.test(translationData.translated_content)) {
      errors.push('Translation contains repeated characters that may indicate encoding issues');
    }

    // Check for unusual character sequences
    if (/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]/.test(translationData.translated_content)) {
      errors.push('Translation contains control characters that may indicate processing errors');
    }
  }

  return {
    isValid: errors.length === 0,
    errors: errors
  };
};

/**
 * Get content with fallback mechanism
 * If translation is not available, return original content
 */
export const getContentWithFallback = async (contentId, originalTitle, originalContent, targetLanguage) => {
  if (targetLanguage === 'en') {
    // For English, always return original content
    return {
      title: originalTitle,
      content: originalContent,
      language: 'en',
      hasTranslation: false
    };
  }

  try {
    // Try to get translation
    const translation = await fetchTranslation(contentId, targetLanguage);

    if (translation && translation.translated_content) {
      // Translation is available
      return {
        title: translation.translated_title || originalTitle,
        content: translation.translated_content,
        language: targetLanguage,
        hasTranslation: true,
        summary: translation.summary
      };
    } else {
      // Translation not available, fallback to original
      console.warn(`Translation not available for content ${contentId} in language ${targetLanguage}, falling back to original`);
      return {
        title: originalTitle,
        content: originalContent,
        language: 'en', // Indicate that we're showing English content
        hasTranslation: false
      };
    }
  } catch (error) {
    console.error(`Error getting translation for ${contentId} in ${targetLanguage}:`, error);
    // On error, fallback to original content
    return {
      title: originalTitle,
      content: originalContent,
      language: 'en',
      hasTranslation: false
    };
  }
};